pkgs/misc/uboot/default.nix

```diff
@@ -179,50 +181,113 @@ in {
+      ${firmwareOdroidC2}/fip_create \
+        --bl30 ${firmwareOdroidC2}/bl30.bin \
+        --bl301 ${firmwareOdroidC2}/bl301.bin \
+        --bl31 ${firmwareOdroidC2}/bl31.bin \
+        --bl33 u-boot.bin \
         fip.bin
-      cat '${firmwareBlobs}/gxb/bl2.package' fip.bin > boot_new.bin
-      '${buildPackages.meson-tools}/bin/amlbootsig' boot_new.bin u-boot.img
-      dd if=u-boot.img of=u-boot.gxbb bs=512 skip=96
+      ${firmwareOdroidC2}/fip_create --dump fip.bin
+
+      cat ${firmwareOdroidC2}/bl2.package fip.bin > boot_new.bin
+      ${firmwareOdroidC2}/aml_encrypt_gxb --bootsig \
+        --input boot_new.bin \
+        --output u-boot.img
+
+      dd if=u-boot.img of=u-boot.bin bs=512 skip=96 status=none
+    '';
+
+    filesToInstall = [
+      "u-boot.bin"
+      "${firmwareOdroidC2}/sd_fusing.sh" "${firmwareOdroidC2}/bl1.bin.hardkernel"
+    ];
+    extraMeta.platforms = ["aarch64-linux"];
+  };
+
+  ubootOdroidC4 = buildUBoot {
+    defconfig = "odroid-c4_defconfig";
+    extraPatches = [
+      # makes C4 board able to boot and use eMMC
+      (fetchpatch {
+        name = "0001-HACK-mmc-meson-gx-limit-to-24MHz.patch";
+        url = "https://gitlab.denx.de/u-boot/custodians/u-boot-amlogic/-/commit/7298cb545aed7aabf5c22b4073d789f1b2e9962b.diff";
+        sha256 = "0ara0zc4pp6bqs42pq8r7dali7ziz6vldw38i0jqadqx8ncg292f";
+      })
+    ];
+
+    postBuild = ''
+      blx_fix() {
+        # BL2 file size 41K, bl21 file size 3K (file size not equal runtime size) = total 44K
+        # after encrypt process, bl2 add 4K header, cut off 4K tail
+        # BL2 file size 56K, acs file size 4K = total 60K
+        # after encrypt process, BL2 add 4K header, cut off 4K tail
+        # BL30 limit 41K
+        # BL301 limit 12K
+        # BL2 limit 56K
+        # ACS limit 4K, but encrypt tool need 64K bl2.bin, so fix to 8192byte.
+
+        case $7 in
+          "bl30")
+            declare blx_bin_limit=40960
+            declare blx01_bin_limit=13312
+          ;;
+          "bl2")
+            declare blx_bin_limit=57344
+            declare blx01_bin_limit=4096
+          ;;
+        esac
+
+        declare -i blx_size=`du -b $1 | awk '{print int($1)}'`
+        declare -i zero_size=$blx_bin_limit-$blx_size
+        dd if=/dev/zero of=$2 bs=1 count=$zero_size
+        cat $1 $2 > $3
+
+        declare -i blx01_size=`du -b $4 | awk '{print int($1)}'`
+        declare -i zero_size_01=$blx01_bin_limit-$blx01_size
+        dd if=/dev/zero of=$2 bs=1 count=$zero_size_01
+        cat $4 $2 > $5
+
+        cat $3 $5 > $6
+      }
+
+      blx_fix ${firmwareOdroidC4}/bl30.bin zero_tmp bl30_zero.bin \
+              ${firmwareOdroidC4}/bl301.bin bl301_zero.bin \
+              bl30_new.bin bl30
+      blx_fix ${firmwareOdroidC4}/bl2.bin zero_tmp bl2_zero.bin \
+              ${firmwareOdroidC4}/acs.bin bl21_zero.bin \
+              bl2_new.bin bl2
+
+      ${firmwareOdroidC4}/aml_encrypt_g12a --bl30sig --input bl30_new.bin \
```

Comment:
Ideally we would have an open source alternative to this tool, but I can't find any. This means native builds will not work.

Reply:
I will keep watching for it.